src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs

```diff
@@ -101,7 +100,14 @@ public TestRequestSender(ProtocolConfig protocolConfig, ITestRuntimeProvider run
 
         // The connectionInfo here is that of RuntimeProvider, so reverse the role of runner.
         _runtimeProvider = runtimeProvider;
-        _communicationEndpoint = communicationEndPoint;
+
+        // TODO: In various places TestRequest sender is instantiated, and we can't easily inject the factory, so this is last
+        // resort of getting the dependency into the execution flow.
+        _communicationEndpoint = communicationEndPoint
+#if DEBUG
```

Comment:
Just to make sure, when in debug we want to have the 3 possible paths, right?

Reply:
Yes, in debug we want to try the service locator before setting the channel.